plugins/regrass.cpp

```diff
@@ -12,109 +27,553 @@ using namespace DFHack;
+
+    if (mat == tiletype_material::GRASS_LIGHT ||
+        mat == tiletype_material::GRASS_DARK ||
+        mat == tiletype_material::PLANT ||
+        tt == tiletype::TreeTrunkPillar)
+    {   // Already appropriate tile
+        DEBUG(log).print("Tiletype no change.\n");
+        return 1;
+    }
+    else if (mat == tiletype_material::STONE ||
+        mat == tiletype_material::LAVA_STONE ||
+        mat == tiletype_material::MINERAL)
+    {   // Muddy non-feature stone
+        for (auto blev : block->block_events)
+        {   // Remove mud spatter
+            if (blev->getType() != block_square_event_type::material_spatter)
+                continue;
+
+            auto &ms_ev = *(df::block_square_event_material_spatterst *)blev;
+            if (ms_ev.mat_type == builtin_mats::MUD)
+            {
+                ms_ev.amount[x][y] = 0;
+                DEBUG(log).print("Removed tile mud.\n");
+                break;
+            }
+        }
+    }
+
+    if (shape == tiletype_shape::FLOOR)
+    {   // Handle random variant, ashes
+        DEBUG(log).print("Tiletype to random grass floor.\n");
+        block->tiletype[x][y] = findRandomVariant((rand() & 1) ? tiletype::GrassLightFloor1 : tiletype::GrassDarkFloor1);
+    }
+    else
+    {
+        auto new_mat = (rand() & 1) ? tiletype_material::GRASS_LIGHT : tiletype_material::GRASS_DARK;
+        auto new_tt = findTileType(shape, new_mat, tiletype_variant::NONE, tiletype_special::NONE, nullptr);
+        DEBUG(log).print("Tiletype to %s.\n", ENUM_KEY_STR(tiletype, new_tt).c_str());
+        block->tiletype[x][y] = new_tt;
+    }
+
+    return 1;
+}
+
+int regrass_block(const regrass_options &options, df::map_block *block)
+{   // Regrass single block
+    CHECK_NULL_POINTER(block);
+
+    int count = 0;
+    for (int x = 0; x < 16; x++)
+    {
+        for (int y = 0; y < 16; y++)
+            count += regrass_tile(options, block, x, y);
+    }
+
+    return count;
+}
+
+int regrass_zlevels(const regrass_options &options, int32_t z1, int32_t z2 = -30000)
+{   // Regrass a range of z-levels
+    if (z1 < 0 || z1 >= world->map.z_count_block)
+    {
+        Core::printerr("z1 out of map bounds!\n");
+        return 0;
+    }
+    else if (z2 != -30000 && (z2 < 0 || z2 >= world->map.z_count_block))
+    {
+        Core::printerr("z2 out of map bounds!\n");
+        return 0;
+    }
+
+    auto z = z2 < 0 ? z1 : std::min(z1, z2);
+    auto z_end = std::max(z1, z2);
+    int count = 0;
+    for (; z <= z_end; z++)
+    {   // Iterate all blocks in z-level range
+        for (int32_t x = 0; x < world->map.x_count_block; x++)
+        {
+            for (int32_t y = 0; y < world->map.y_count_block; y++)
+            {
+                auto block = Maps::getBlock(x, y, z);
+                if (block)
+                    count += regrass_block(options, block);
+                else // Probably below HFS
+                    Core::print("No getBlock(%d, %d, %d)! Skipping.\n", x, y, z);
```

Comment:
usage of `Core::print` will cause output to fail to be captured by streams (e.g. `gui/launcher`). Normally, plugins will pass the `color_ostream & out` output sink to all the internal functions for debug output. This goes for the DEBUG statements as well. The `out` param should be passed to those as well so their output can get captured.